parquet-column/src/main/java/org/apache/parquet/io/RecordReaderImplementation.java

```diff
@@ -424,13 +434,105 @@ public T read() {
     return record;
   }
 
+  private State getState(MessageType schema) {
+    for (State state : states) {
+        if (schema.containsField(state.fieldPath[0])) {
+          return state;
+        }
+    }
+    throw new IllegalArgumentException("Couldn't find state for schema " + schema);
+  }
+
   @Override
-  public boolean shouldSkipCurrentRecord() {
-    return shouldSkipCurrentRecord;
+  public void readVectors(ColumnVector[] vectors, MessageType[] columnSchemas, long current, long total) {
+      for (int i = 0 ; i < columnSchemas.length; i++) {
+        MessageType schema  = columnSchemas[i];
+        State state = getState(schema);
+        readVector(state.column, vectors[i], current, total);
+      }
   }
 
-  private static void log(String string) {
-    LOG.debug(string);
+  @Override
+  public void readVector(ObjectColumnVector<T> vector, long current, long total) {
+    int index = 0;
+    for ( ; index < ColumnVector.DEFAULT_VECTOR_LENGTH; index++, current++) {
+
+      if (current >= total) {
+        ((ColumnVector) vector).setNumberOfValues(index);
+        return;
+      }
+
+      T record = read();
+
+      if (shouldSkipCurrentRecord) {
+        if (DEBUG) {
+          LOG.debug("skipping record");
+        }
+        vector.isNull[index] = true;
+      }
+      vector.values[index] = record;
+    }
+
+    ((ColumnVector) vector).setNumberOfValues(index);
+  }
+
+  private void readVector(ColumnReader reader, ColumnVector vector, long current, long total) {
+    ColumnDescriptor column = reader.getDescriptor();
+    int maxDefinitionLevel = column.getMaxDefinitionLevel();
+
+    int index = 0;
+    int fixedLenByteArrayPosition = 0;
+    for ( ; index < ColumnVector.DEFAULT_VECTOR_LENGTH; index++, current++) {
+
+      if (current >= total) {
+        vector.setNumberOfValues(index);
+        return;
+      }
+
+      if (reader.getCurrentDefinitionLevel() == maxDefinitionLevel) {
+        switch (column.getType()) {
```

Comment:
Added a `readFrom` method, but we still need the switch/case to handle INT96 and FIXED_LEN_BYTE_ARRAY in a special way. The switch statement is simplified anyway.